rj_gameplay/stp/play/__init__.py

```diff
@@ -87,41 +90,45 @@ def assign_roles(
 
         used_robots = set()
         for tactic in self.prioritized_tactics:
-            # TODO: handle if tactic requests more roles than exist
-            #       by not assigning tactic at all!
-            #       and give some default behavior (later)
             robots_for_tactic = []
             for cost_fn, role in tactic.role_requests:
                 min_cost = 1e9
                 cheapest_robot = None
 
                 for robot in world_state.our_robots:
-                    if robot in used_robots or not robot.visible:
+                    if robot in used_robots or robot in robots_for_tactic:
+                        continue
+                    if not robot.visible:
                         continue
                     cost = cost_fn(robot, world_state)
                     if cost < min_cost:
                         min_cost = cost
                         cheapest_robot = robot
 
                 if cheapest_robot is None:
-                    # TODO: consider looping until success?
-                    raise RoleAssignFailure(
-                        f"RoleRequest ({role}, {cost_fn}) was not assigned"
-                    )
-
-                used_robots.add(cheapest_robot)
+                    print(f"RoleRequest ({role}, {cost_fn}) was not assigned")
+                    robots_for_tactic = None
+                    break
                 robots_for_tactic.append(cheapest_robot)
-            tactic.set_assigned_robots(robots_for_tactic)
 
-        for tactic in self.prioritized_tactics:
-            tactic.init_roles(world_state)
+            if robots_for_tactic is not None:
+                used_robots.update(robots_for_tactic)
+                tactic.set_assigned_robots(robots_for_tactic)
+                self.approved_prioritized_tactics.append(tactic)
+                tactic.init_roles(world_state)
+            else:
+                print("Tactic denied")
```

Comment:
Would be helpful if this printed something like "Tactic <name> denied: requested <x> roles, but only <y> robots available" or something of that nature